Wikipedia/Code/DiffContainerViewController.swift

```diff
@@ -0,0 +1,495 @@
+            return
+        }
+        
+        let singleDiffVC = DiffContainerViewController(articleTitle: articleTitle, siteURL: siteURL, type: .single(byteDifference: nextRevision.revisionSize), fromModel: toModel, toModel: nextRevision, theme: theme, revisionDelegate: revisionDelegate)
+        wmf_push(singleDiffVC, animated: true)
+    }
+    
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        
+        switch type {
+        case .compare:
+            self.showDiffPanelOnce()
+        case .single:
+            break
+        }
+    }
+    
+    override func viewDidLayoutSubviews() {
+        super.viewDidLayoutSubviews()
+        
+        if let scrollView = diffListViewController?.scrollView {
+            configureExtendedViewSquishing(scrollView: scrollView)
+        }
+
+        if let emptyViewController = scrollingEmptyViewController {
+            navigationBar.setNeedsLayout()
+            navigationBar.layoutSubviews()
+            let bottomSafeAreaHeight = view.bounds.height - safeAreaBottomAlignView.frame.maxY
+            let targetRect = CGRect(x: 0, y: navigationBar.visibleHeight, width: emptyViewController.view.frame.width, height: emptyViewController.view.frame.height - navigationBar.visibleHeight - bottomSafeAreaHeight)
+            //tonitodo: this still doesn't seem quite centered...
+            let convertedTargetRect = view.convert(targetRect, to: emptyViewController.view)
+            print(convertedTargetRect)
+            emptyViewController.centerEmptyView(within: convertedTargetRect)
+        }
+    }
+    
+    override func apply(theme: Theme) {
+        
+        guard isViewLoaded else {
+            return
+        }
+        
+        super.apply(theme: theme)
+        
+        view.backgroundColor = theme.colors.paperBackground
+        
+        headerTitleView?.apply(theme: theme)
+        headerExtendedView?.apply(theme: theme)
+        diffListViewController?.apply(theme: theme)
+        scrollingEmptyViewController?.apply(theme: theme)
+    }
+}
+
+private extension DiffContainerViewController {
+    
+    func evaluateState() {
+        
+        switch containerViewModel.state {
+
+        case .loading:
+            fakeProgressController.start()
+            scrollingEmptyViewController?.view.isHidden = true
+            diffListViewController?.view.isHidden = true
+            wmf_hideEmptyView()
+        case .empty:
+            fakeProgressController.stop()
+            setupScrollingEmptyViewControllerIfNeeded()
+            switch type {
+            case .compare:
+                scrollingEmptyViewController?.type = .diffCompare
+            case .single:
+                scrollingEmptyViewController?.type = .diffSingle
+            }
+            scrollingEmptyViewController?.view.isHidden = false
+            diffListViewController?.view.isHidden = true
+            wmf_hideEmptyView()
```

Comment:
for the `empty` state, shouldn't this trigger showing the empty view via a call to `wmf_showEmptyViewOfType`?

Reply:
Good catch, I removed these hide calls, it was old code from a different attempt. I'm using EmptyViewController for all of it - I created it for pull to refresh on empty talk pages and now leaning on it for diff. It's headers get so large with dynamic type that the empty states are cut off without scrolling.